### PR TITLE
Fix OpenSearch Retry mechanism

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
Fixes Open Search retry logic
Modified to use com.linecorp.armeria.client.retry.Backoff for back off policy. This policy uses initial delay, max delay and max retries. The first retry is done after initial delay. And exponential delay is done until max delay is reached. After that it uses max delay for subsequent retry attempts. Total max retries number of attempts are made.

Initial delay is set to 50ms, max delay is set to 10 minutes. These two values are currently not configurable. Max retries is configurable.

With these values, 10 minutes is reached after 15 iterations. Suggested max_retries is 25 or 30.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [X ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java [hunk 1]
```diff
@@ -184,20 +182,28 @@ private void incrementErrorCounters(final Exception e) {
     }
 
     public void execute(final AccumulatingBulkRequest bulkRequest) throws InterruptedException {
-        // Exponential backoff run forever
-        // TODO: replace with custom backoff policy setting including maximum interval between retries
-        final BackOffUtils backOffUtils = new BackOffUtils(
-                BackoffPolicy.exponentialBackoff(TimeValue.timeValueMillis(50), Integer.MAX_VALUE).iterator());
+        final Backoff backoff = Backoff.exponential(INITIAL_DELAY_MS, MAXIMUM_DELAY_MS).withMaxAttempts(maxRetries);
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java:185]: The original behavior of the sink was to have unlimited retries. Should this have some sort of configuration such as the following:

```
Backoff backoff = Backoff.exponential(INITIAL_DELAY_MS, MAXIMUM_DELAY_MS)
if(maxRetries != 0) {
  backoff = backoff.withMaxAttempts(maxRetries);
}
```

File: data-prepper-plugins/opensearch/build.gradle [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 dependencies {
     implementation project(':data-prepper-api')
+    implementation libs.armeria.core
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/build.gradle:8]: Thanks for using this!

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java [hunk 3]
```diff
@@ -184,20 +182,28 @@ private void incrementErrorCounters(final Exception e) {
     }
 
     public void execute(final AccumulatingBulkRequest bulkRequest) throws InterruptedException {
-        // Exponential backoff run forever
-        // TODO: replace with custom backoff policy setting including maximum interval between retries
-        final BackOffUtils backOffUtils = new BackOffUtils(
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java:189]: Can we now remove `BackOffUtils`? Is it being used elsewhere?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java [hunk 4]
```diff
@@ -184,20 +182,28 @@ private void incrementErrorCounters(final Exception e) {
...
-            if (operationResponse != null && backOffUtils.hasNext()) {
-                // Wait for backOff duration
-                backOffUtils.next();
+            operationResponse = handleRetry(request, response, attempt);
+            if (operationResponse != null) {
+                final long delayMillis = backoff.nextDelayMillis(attempt++);
                 request = operationResponse.getBulkRequest();
                 response = operationResponse.getResponse();
+                if (delayMillis < 0) {
+                    RuntimeException e = new RuntimeException(String.format("Number of retries reached the limit of max retries(configured value %d)", maxRetries));
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java:197]: We should have a space in between "max retries" and "(configured value)"

> ...max retries (configured value...

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java:185]: It appears that the current configuration uses `Integer.MAX_VALUE` to indicate infinite retries.

https://github.com/opensearch-project/data-prepper/blob/738293eb4f182764d2f4e7e3ed2e686ca973cab7/data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/RetryConfiguration.java#L43

I think we can hook off of this for this PR. But, we should probably consider a better approach in the future. As it is, you cannot explicitly set infinite retries in the YAML without knowing the maximum value in Java (which is rather opaque for a user in my opinion).

Comment (kkondaka) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java:185]: max retries is always set to default value of `Integer.MAX_VALUE`

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java (hunk 3 above)
Comment (kkondaka) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java:189]: Removed BackOffUtils.

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java:185]: The `max_retries` can be set by the user.

What I'm suggesting for this PR is that we don't create a very large limit, but actually make it unlimited. By not calling `withMaxAttempts(int)` there will be no maximum at all.

Backoff backoff = Backoff.exponential(INITIAL_DELAY_MS, MAXIMUM_DELAY_MS)
if(maxRetries != Integer.MAX_VALUE) {
  backoff = backoff.withMaxAttempts(maxRetries);
}

**Review by dlvenable — APPROVED** [review]

**Review by oeyh — APPROVED** [review]

**Review by asifsmohammed — APPROVED** [review]
LGTM!
We should make the max retries configurable in S3 source as well.

**Comment by dlvenable** [comment]
This PR fixes #2641

**Pull request merged**